haystack/preprocessor/preprocessor.py

```diff
@@ -0,0 +1,205 @@
+import logging
+import re
+from copy import deepcopy
+from functools import partial, reduce
+from itertools import chain
+from typing import List, Optional, Tuple, Generator, Set
+
+import nltk
+from more_itertools import windowed
+
+from haystack.preprocessor.base import BasePreProcessor
+
+logger = logging.getLogger(__name__)
+
+
+class PreProcessor(BasePreProcessor):
+    def __init__(
+        self,
+        clean_whitespace: Optional[bool] = True,
+        clean_header_footer: Optional[bool] = False,
+        clean_empty_lines: Optional[bool] = True,
+        split_by: Optional[str] = "passage",
+        split_size: Optional[int] = 10,
+        split_stride: Optional[int] = None,
+        split_mid_sentence: Optional[bool] = True,
+    ):
+        """
+        :param clean_header_footer: use heuristic to remove footers and headers across different pages by searching
+                                     for the longest common string. This heuristic uses exact matches and therefore
+                                     works well for footers like "Copyright 2019 by XXX", but won't detect "Page 3 of 4"
+                                     or similar.
+        :param clean_whitespace: strip whitespaces before or after each line in the text.
+        :param clean_empty_lines: remove more than two empty lines in the text.
+        :param split_by: split the document by "word", "sentence", or "passage". Set to None to disable splitting.
```

Comment:
what do you think about adding token and word ?
word splitted by space and token generated from tokenizer ?

Reply:
We started with whitespace-based "words" here as it's faster, simpler, doesn't depend on a model's tokenizer and is quite popular these days in research (see DPR paper). 
I believe it would make sense to add the option for "tokens" here later, too. However, this would mean that preprocessing depends on the reader model and if I switch between models I cannot rely on meaningful splits anymore and would need to re-index everything. This should be clearly communicated and designed here.